crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -350,19 +352,39 @@ where
 
     /// Create a new package
     pub fn new_package(
-        &mut self,
+        &self,
         modules: Vec<move_binary_format::CompiledModule>,
-    ) -> Result<ObjectID, ExecutionError> {
-        // wrap the modules in an object, write it to the store
-        let package_object = Object::new_package(
+        version: Option<SequenceNumber>,
+    ) -> Result<Object, ExecutionError> {
+        Object::new_package(
             modules,
-            OBJECT_START_VERSION,
+            version.unwrap_or(OBJECT_START_VERSION),
             self.tx_context.digest(),
             self.protocol_config.max_move_package_size(),
-        )?;
-        let id = package_object.id();
-        self.new_packages.push(package_object);
-        Ok(id)
+        )
+    }
+
+    /// Upgrades the package: checks compatibility and increments the version number.
+    pub fn upgrade_package<'x, D: IntoIterator<Item = &'x MovePackage>>(
+        &self,
+        previous_package: &MovePackage,
+        new_modules: Vec<move_binary_format::CompiledModule>,
+        dependencies: D,
+        compatiblity_checker: Compatibility,
+    ) -> Result<Object, ExecutionError> {
+        Object::new_upgraded_package(
+            previous_package,
+            new_modules,
+            self.tx_context.digest(),
+            self.protocol_config.max_move_package_size(),
+            dependencies,
+            compatiblity_checker,
```

Comment:
Note that eventually the set of compatibility checks will extend beyond what the `Compatibility` type offers (i.e. `ADDITIVE` and `DEP_ONLY` compatibility).  So it may be worth handling compatibility checks entirely outside the object creation flow.